### ISSUE TITLE
Schema Type Bug with Arrays & Nested Types Inside Array

### ISSUE THREAD

**Issue by armenr** [roboncode/orango#73]
Greetings - I'm working with a schema, and at one of the document properties, I'd like to have an array, in which are nested n (or zero) number of objects. 

How do we get this to work? Currently, I am getting the following error:

```
because [child "images" fails because ["images" must be an object]
```

My Schema:
```
const SomeUserDataModel = (orango) => {
  // const { OPERATIONS } = orango.consts;
  let schema = new orango.Schema({
    birthdate: String,
    country: String,
    images: [Object],
    product: String,
    someKindofListType: String,
  });

  orango.model('someuserdata', schema);
};

export default SomeUserDataModel;
```

This might be a headslappingly stupid question, but how do I get this to work? :)

**Comment by armenr** [comment]
And while we're at it, could ask this as well?

Does the ODM provide any sort of inbuilt functionality that says: "Go look up a user by this e-mail address/username, and IF they exist, do not allow another document to be written with these same properties", or is that a thing we should just code up ourselves?

**Comment by roboncode** [comment]
For question #1 - I am guessing here, but I believe you may be trying to do this....
images: [{}]

Let me know if that fixes your issue.

**Comment by armenr** [comment]
Testing now. Will update. Thank you!

**Comment by armenr** [comment]
Doesn't seem to work. Can't tell if it's an issue with Tang, or if it's me doing something wrong. See below:

```
const SomeUserDataModel = (orango) => {
  // const { OPERATIONS } = orango.consts;
  let schema = new orango.Schema({
    birthdate: String,
    country: String,
    images: [{}],
    product: String,
    someKindofListType: String,
  });

  orango.model('someuserdata', schema);
};

export default SomeUserDataModel;
```

```
PAYLOAD TO INSERT! {
<REDACTED>
  id: 'myID',
  images: [
    {
      height: null,
      url: 'https://some.url/somefile.jpg',
      width: null
    }
  ],
  someOtherKey: 'someOtherValue'
}
(node:89495) UnhandledPromiseRejectionWarning: ValidationError: child "accounts" fails because [child "someAccountData" fails because [child "images" fails because ["images" must be an object]]]
    at Object.exports.process (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/joi/lib/errors.js:201:19)
    at internals.Object._validateWithOptions (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/joi/lib/types/any/index.js:751:31)
    at internals.Object.validate (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/joi/lib/types/any/index.js:785:21)
    at Schema.validate (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Schema.js:74:16)
    at Schema.validate (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/Schema.js:111:28)
    at OrangoModel.validate (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Model.js:92:33)
    at OrangoModel.toObject (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Model.js:115:25)
    at Builder.<anonymous> (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/builder.js:18:22)
    at /Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/builder.js:64:39
    at asyncForEach (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/helpers/asyncForEach.js:11:18)
(node:89495) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:89495) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
ServiceUnavailableError: Response timeout
    at IncomingMessage.<anonymous> (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/connect-timeout/index.js:84:8)
    at IncomingMessage.emit (events.js:196:13)
    at IncomingMessage.EventEmitter.emit (domain.js:471:20)
    at Timeout._onTimeout (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/connect-timeout/index.js:49:11)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)
```

**Comment by armenr** [comment]
See this snippet, where I attempt to use Joi instead

```
let schema = new orango.Schema({
    images: Joi.array(),
```

```
[myproject-auth] [nodemon] starting `babel-node ./src/server.js`
[myproject-auth] RangeError: Maximum call stack size exceeded
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:1:35)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
[myproject-auth]     at deepCopy (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/helpers/deepCopy.js:6:20)
```

**Comment by armenr** [comment]
Some more peculiarities when I log the schema that's being returned from model:

```
[myproject-auth]     types: {
[myproject-auth]       String: [Function: String],
[myproject-auth]       Number: [Function: Number],
[myproject-auth]       Boolean: [Function: Boolean],
[myproject-auth]       Object: [Function: Object],
[myproject-auth]       Array: [Object],
[myproject-auth]       Date: [Function: Date],
[myproject-auth]       RegExp: [Function: RegExp],
[myproject-auth]       Id: [Object],
[myproject-auth]       Any: [Object],
[myproject-auth]       Mixed: [Object]
[myproject-auth]     }
```

When I try `images: Any`, I get:

```
[myproject-auth] ReferenceError: Any is not defined
[myproject-auth]     at someUserDataModel (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/models/someUserData.js:13:13)
[myproject-auth]     at connectArangoDB (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/connectors/connectArango.js:14:11)
[myproject-auth]     at start (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/server.js:29:31)
[myproject-auth]     at Object.<anonymous> (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/server.js:62:1)
[myproject-auth]     at Module._compile (internal/modules/cjs/loader.js:759:30)
[myproject-auth]     at loader (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/babel-register/lib/node.js:144:5)
[myproject-auth]     at Object.require.extensions.<computed> [as .js] (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/babel-register/lib/node.js:154:7)
[myproject-auth]     at Module.load (internal/modules/cjs/loader.js:628:32)
[myproject-auth]     at Function.Module._load (internal/modules/cjs/loader.js:555:12)
[myproject-auth]     at Function.Module.runMain (internal/modules/cjs/loader.js:822:10)
```
When I try `    images: { type: Array }` I get:

```
myproject-auth] TypeError: Cannot read property 'schema' of undefined
[myproject-auth]     at Schema._parse (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Schema.js:84:14)
[myproject-auth]     at Schema._parse (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Schema.js:95:30)
[myproject-auth]     at Schema._parse (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Schema.js:115:31)
[myproject-auth]     at new Schema (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/tangjs/lib/Schema.js:27:22)
[myproject-auth]     at Schema.init (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/Schema.js:26:23)
[myproject-auth]     at Orango.createModel (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/Orango.js:275:12)
[myproject-auth]     at Orango.model (/Users/armenr/Desktop/myproject-new/myprojectzer0/node_modules/orango/lib/Orango.js:357:35)
[myproject-auth]     at someUserDataModel (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/models/someUserData.js:21:10)
[myproject-auth]     at connectArangoDB (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/connectors/connectArango.js:14:11)
[myproject-auth]     at start (/Users/armenr/Desktop/myproject-new/myprojectzer0/packages/myproject-auth/src/server.js:29:31)
```

**Comment by roboncode** [comment]
Thx for the thorough example, I am looking into it.

**Comment by armenr** [comment]
Sure thing! Appreciate your help :)

**Comment by armenr** [comment]
@roboncode - any hints or tips? I've made some other small fixes and created two PR's, but this problem is a bit prohibitive for us :)

If you're tied up, we'll try to investigate or fix it, but any hints or help or tips would be appreciated if you're unable to investigate/bugfix.

**Comment by roboncode** [comment]
This has been fixed with the new release of v1.0.0-beta-1.1
